components/tidb_query_vec_expr/src/impl_compare.rs

```diff
@@ -268,7 +359,39 @@ pub fn greatest_time(ctx: &mut EvalContext, args: &[Option<BytesRef>]) -> Result
         }
     }
 
-    Ok(greatest.map(|time| time.to_string().into_bytes()))
+    if invalid_time != None {
+        return Ok(invalid_time);
```

Comment:
I'd like to know the reason for the special handling here. 
What I tried in MySQL 8.0.20:
```
mysql root@localhost:(none)> select least(date("2020-01-01"), date("2020-01-00"))
+-------------------------------------------------+
| least(date("2020-01-01"), date("2020-01-00"))   |
|-------------------------------------------------|
| 2020-01-00                                      |
+-------------------------------------------------+
1 row in set
Time: 0.002s
mysql root@localhost:(none)> select least(date("2020-01-01"), date("2020-01-02"))
+-------------------------------------------------+
| least(date("2020-01-01"), date("2020-01-02"))   |
|-------------------------------------------------|
| 2020-01-01                                      |
+-------------------------------------------------+
1 row in set
Time: 0.002s
mysql root@localhost:(none)> select least(date("2020-01-01"), date("invalid_time"))
+---------------------------------------------------+
|   least(date("2020-01-01"), date("invalid_time")) |
|---------------------------------------------------|
|                                            <null> |
+---------------------------------------------------+
```

Reply:
In TiDB, both non-vec and vec LeastTime functions will make the first invalid time as result.

https://github.com/pingcap/tidb/blob/master/expression/builtin_compare_vec.go#L763
https://github.com/pingcap/tidb/blob/master/expression/builtin_compare.go#L815

And the reason your MySQL query returns NULL is because `date()` will convert an invalid time format to NULL.

```sql
mysql> select date("invalid_time");
+----------------------+
| date("invalid_time") |
+----------------------+
| NULL                 |
+----------------------+
1 row in set, 1 warning (0.00 sec)
```

Reply:
I'm assuming that when TiDB and MySQL handle the Least() function, they will chose a certain type Least() such as LeastTime() to compare all arguments which might not all be DATETIME. In that case, Least() will return the first invalid time it found.

```sql
mysql>   SELECT '21:50:04.003414' NOT LIKE `col_bigint_unsigned_key` AS Result, `col_char_255` AS col_char_255, `col_date_key` AS col_date_key, CAST( `col_date_key` AS DATETIME ) AS col_date_key_datetime, LEAST( `col_date_key`,`col_char_255` ) AS LeastResult FROM `table1000_int_autoinc` WHERE LEAST( `col_date_key`,`col_char_255` ) ORDER BY LEAST( `col_date_key`,`col_char_255` ) LIMIT 20; 
+--------+--------------+--------------+-----------------------+-------------+
| Result | col_char_255 | col_date_key | col_date_key_datetime | LeastResult |
+--------+--------------+--------------+-----------------------+-------------+
|      1 | -1141833728  | 2007-11-18   | 2007-11-18 00:00:00   | -1141833728 |
|      1 | -1188167680  | 2007-06-27   | 2007-06-27 00:00:00   | -1188167680 |
|      1 | -1310588928  | 2008-02-14   | 2008-02-14 00:00:00   | -1310588928 |
|      1 | -1333067776  | 0000-00-00   | 0000-00-00 00:00:00   | -1333067776 |
|      1 | -1338179584  | 0000-00-00   | 0000-00-00 00:00:00   | -1338179584 |
|      1 | -1437138944  | 0000-00-00   | 0000-00-00 00:00:00   | -1437138944 |
|      1 | -1461780480  | 2001-11-27   | 2001-11-27 00:00:00   | -1461780480 |
|      1 | -1514012672  | 2005-03-03   | 2005-03-03 00:00:00   | -1514012672 |
|      1 | -1567817728  | 0000-00-00   | 0000-00-00 00:00:00   | -1567817728 |
|      1 | -156958720   | 0000-00-00   | 0000-00-00 00:00:00   | -156958720  |
|   NULL | -1588396032  | 2000-07-07   | 2000-07-07 00:00:00   | -1588396032 |
|      1 | -1693188096  | 2009-11-20   | 2009-11-20 00:00:00   | -1693188096 |
|      1 | -1739390976  | 0000-00-00   | 0000-00-00 00:00:00   | -1739390976 |
|      1 | -1765736448  | 0000-00-00   | 0000-00-00 00:00:00   | -1765736448 |
|      1 | -1768554496  | 0000-00-00   | 0000-00-00 00:00:00   | -1768554496 |
|      1 | -1789526016  | 0000-00-00   | 0000-00-00 00:00:00   | -1789526016 |
|      1 | -1877409792  | 0000-00-00   | 0000-00-00 00:00:00   | -1877409792 |
|      1 | -1878982656  | 0000-00-00   | 0000-00-00 00:00:00   | -1878982656 |
|      1 | -1910308864  | 0000-00-00   | 0000-00-00 00:00:00   | -1910308864 |
|      1 | -1937440768  | 0000-00-00   | 0000-00-00 00:00:00   | -1937440768 |
+--------+--------------+--------------+-----------------------+-------------+
20 rows in set, 1294 warnings (0.04 sec)
```

Reply:
A little strange to me. :\

However, another incompatible stuff found.
In MySQL:
```
> create table t (a datetime, b int);
> insert into t values ("2020-01-01", 2020202020)
> select least(a, b) from t;
+---------------------+
| least(a, b)         |
|---------------------|
| 2020-01-01 00:00:00 |
+---------------------+
```

In TiDB:
```
> create table t (a datetime, b int);
> insert into t values ("2020-01-01", 2020202020)
> select least(a, b) from t;
+---------------+
|   least(a, b) |
|---------------|
|    2020202020 |
+---------------+
```

Reply:
```sql
drop table t;
create table t (int_type int, str_type char(255), datetime_type datetime);
insert into t values (12345, "invalid_time", "2020-01-01");
select int_type, str_type, datetime_type from t;
select least(int_type, str_type, datetime_type) from t;
select least(int_type, str_type) from t;
select least(int_type, datetime_type) from t;
select least(str_type, datetime_type) from t;
```

I tried these sqls. Here are results.

In MySQL:

```sql
mysql> source /Users/jmpotato/test.sql
Query OK, 0 rows affected (0.01 sec)

Query OK, 0 rows affected (0.01 sec)

Query OK, 1 row affected (0.00 sec)

+----------+--------------+---------------------+
| int_type | str_type     | datetime_type       |
+----------+--------------+---------------------+
|    12345 | invalid_time | 2020-01-01 00:00:00 |
+----------+--------------+---------------------+
1 row in set (0.00 sec)

+------------------------------------------+
| least(int_type, str_type, datetime_type) |
+------------------------------------------+
| 12345                                    |
+------------------------------------------+
1 row in set, 2 warnings (0.00 sec)

Warning (Code 1292): Incorrect datetime value: '12345' for column 'datetime_type' at row 1
Warning (Code 1292): Incorrect datetime value: 'invalid_time' for column 'datetime_type' at row 1
+---------------------------+
| least(int_type, str_type) |
+---------------------------+
| 12345                     |
+---------------------------+
1 row in set (0.01 sec)

+--------------------------------+
| least(int_type, datetime_type) |
+--------------------------------+
| 12345                          |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Incorrect datetime value: '12345' for column 'datetime_type' at row 1
+--------------------------------+
| least(str_type, datetime_type) |
+--------------------------------+
| 2020-01-01 00:00:00            |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Incorrect datetime value: 'invalid_time' for column 'datetime_type' at row 1
```

In TiDB:

```sql
mysql> source /Users/jmpotato/test.sql
ERROR 1051 (42S02): Unknown table 'test.t'
Query OK, 0 rows affected (0.01 sec)

Query OK, 1 row affected (0.01 sec)

+----------+--------------+---------------------+
| int_type | str_type     | datetime_type       |
+----------+--------------+---------------------+
|    12345 | invalid_time | 2020-01-01 00:00:00 |
+----------+--------------+---------------------+
1 row in set (0.00 sec)

+------------------------------------------+
| least(int_type, str_type, datetime_type) |
+------------------------------------------+
|                                        0 |
+------------------------------------------+
1 row in set, 1 warning (0.00 sec)

+---------------------------+
| least(int_type, str_type) |
+---------------------------+
|                         0 |
+---------------------------+
1 row in set, 1 warning (0.01 sec)

+--------------------------------+
| least(int_type, datetime_type) |
+--------------------------------+
|                          12345 |
+--------------------------------+
1 row in set (0.00 sec)

+--------------------------------+
| least(str_type, datetime_type) |
+--------------------------------+
| invalid_time                   |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)
```

It seems that TiDB still has some incompatible bugs...I will keep on investigating to fix this.

Reply:
Glad to see it! TiDB is getting better because of you. XD

Reply:
MAKE TIDB GREAT AGAIN!!!